### PR TITLE
Add Users support

### PR DESCRIPTION
**Pull request by luismfonseca**
Adding Users support, currently missing the Publish to users support.

### CODE REVIEW

File: push_notification_test.go [hunk 1]
```diff
@@ -31,36 +33,36 @@ func TestPushNotifications(t *testing.T) {
...
+			So(err.Error(), ShouldContainSubstring, "Secret Key cannot be an empty string")
 			So(noPN, ShouldBeNil)
 		})
 
 		pn, noErrors := New(testInstanceId, testSecretKey)
 		So(noErrors, ShouldBeNil)
 		So(pn, ShouldNotBeNil)
 
-		Convey("when publishing", func() {
+		Convey("when publishing interests", func() {
```
Comment (luismfonseca) [inline push_notification_test.go]: `pushing to interests`?

File: push_notification_test.go [hunk 2]
```diff
@@ -152,5 +170,197 @@ func TestPushNotifications(t *testing.T) {
...
+
+			Convey("should return an error if the User Id is too long", func() {
+				s := ""
+				for i := 0; i < maxUserIdLength; i++ {
+					s += "a"
+				}
+
+				token, err := pn.AuthenticateUser(s)
+
+				So(err, ShouldBeNil)
```
Comment (luismfonseca) [inline push_notification_test.go]: Pretty good, checking the boundaries! :)

File: push_notification_test.go [hunk 3]
```diff
@@ -152,5 +170,197 @@ func TestPushNotifications(t *testing.T) {
...
+					fmt.Sprintf("User Id ('%s') length too long (expected fewer than %d characters, got %d)", longerUserId, maxUserIdLength+1, len(longerUserId)),
+				)
+			})
+
+			Convey("should return a valid JWT token if everything is correct", func() {
+				token, err := pn.AuthenticateUser("u-123")
+				expectedIssuer := "https://" + testInstanceId + ".pushnotifications.pusher.com"
+				expectedSubject := "u-123"
+
+				So(err, ShouldBeNil)
```
Comment (luismfonseca) [inline push_notification_test.go]: I'd have these `err shouldBeNil` always right after calling a function as to not forget it about it.

I'd move these `expected` strings to where they are being used (principle of smallest variable scope).

File: push_notification_test.go [hunk 4]
```diff
@@ -152,5 +170,197 @@ func TestPushNotifications(t *testing.T) {
...
+				So(parsedToken.Claims.(jwt.MapClaims)["sub"], ShouldEqual, expectedSubject)
+				expirySeconds := parsedToken.Claims.(jwt.MapClaims)["exp"]
+				expiry := expirySeconds.(float64)
+				So(time.Unix(int64(expiry), 0), ShouldHappenAfter, time.Now())
+			})
+		})
+
+		Convey("when publishing to Users", func() {
+			Convey("should fail if no Users are given", func() {
+
```
Comment (luismfonseca) [inline push_notification_test.go]:
```suggestion

```

File: push_notifications.go [hunk 5]
```diff
@@ -8,20 +8,32 @@ import (
...
+	// Creates a signed JWT for a user id.
+	// Returns a signed JWT if successful, or a non-nil `error` otherwise.
+	AuthenticateUser(userId string) (string, error)
 }
 
 const (
 	defaultRequestTimeout     = time.Minute
 	defaultBaseEndpointFormat = "https://%s.pushnotifications.pusher.com"
+	maxUserIdLength           = 164
+	maxNumUserIds             = 1000
```
Comment (luismfonseca) [inline push_notifications.go]:
```
maxNumUserIdsWhenPublishing = 1000
```

File: push_notifications.go [hunk 6]
```diff
@@ -8,20 +8,33 @@ import (
...
 	"github.com/pkg/errors"
 )
 
 // The Pusher Push Notifications Server API client
 type PushNotifications interface {
 	// Publishes notifications to all devices subscribed to at least 1 of the interests given
-	// Returns a non-empty `publishId` string if successful; or a non-nil `error` otherwise.
+	// Returns a non-empty `publishId` JSON string if successful; or a non-nil `error` otherwise.
+	PublishToInterests(interests []string, request map[string]interface{}) (publishId string, err error)
+	// An alias for `PublishToInterests`
```
Comment (luismfonseca) [inline push_notifications.go]:
```
Deprecated: Use `PublishToInterests` instead.
```

Following this: https://rakyll.org/deprecated/. GoLand supports these deprecated comments, for example.

File: push_notifications.go [hunk 7]
```diff
@@ -8,20 +8,33 @@ import (
...
 }
 
 const (
-	defaultRequestTimeout     = time.Minute
-	defaultBaseEndpointFormat = "https://%s.pushnotifications.pusher.com"
+	defaultRequestTimeout       = time.Minute
+	defaultBaseEndpointFormat   = "https://%s.pushnotifications.pusher.com"
+	maxUserIdLength             = 164
+	maxNumUserIdsWhenPublishing = 1000
+	tokenTTL                    = 24
```
Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

Comment (luismfonseca) [inline push_notifications.go]: This should be a `time.Duration`

File: push_notifications.go (hunk 6 above)
Comment (luismfonseca) [inline push_notifications.go]:
```
Deprecated: Use `PublishToInterests` instead.
```

Following this: https://rakyll.org/deprecated/. GoLand supports these deprecated comments, for example.

**Comment by minaorangina** [comment]
Pending public feature release

**Comment by lukabratos** [comment]
Might be also worth updating the [examples](https://github.com/pusher/push-notifications-go/blob/master/example/example.go) code. Currently we only have a `publish` request.

**Pull request merged**